src/libopensc/card-piv.c

```diff
@@ -2635,6 +2681,92 @@ static int piv_process_discovery(sc_card_t *card)
 	LOG_FUNC_RETURN(card->ctx, r);
 }
 
+/*
+ * parse a CCC to test  if this is a Dual CAC/PIV
+ * We read teh CCC using the PIV API.
+ * Look for CAC RID=A0 00 00 00 79
+ */
+ static int piv_parse_ccc(sc_card_t *card, u8* rbuf, size_t rbuflen)
+{
+	int r = 0;
+	const u8 * body;
+	size_t bodylen;
+
+	const u8 *p;
+	const u8 *next;
+	const u8 *end;
+	unsigned int cla_out, tag_out;
+
+	piv_private_data_t * priv = PIV_DATA(card);
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	if (rbuf == NULL || rbuflen == 0) {
+		r = SC_ERROR_WRONG_LENGTH;
+		goto  err;
+	}
+
+	body = rbuf;
+	if ((r = sc_asn1_read_tag(&body, rbuflen, &cla_out, &tag_out,  &bodylen)) != SC_SUCCESS) {
+		sc_log(card->ctx, "DER problem %d",r);
+		r = SC_ERROR_INVALID_ASN1_OBJECT;
+		goto err;
+	}
+
+	priv->ccc_flags |= PIV_CCC_FOUND;
+
+	/* CCC boby has 1 byte tags, and 1 byte lengths */
+	p = body;
+	end = p + bodylen;
+	while (p <= (end - 2) && (next = (p + *(p + 1) + 2)) <= end ) {
```

Comment:
piv_parse_ccc was rewritten to use sc_simpletlv_read_tag

Reply:
Note: there is a cast in the code:
```
r = sc_simpletlv_read_tag((u8**)&body, end - body , &tag, &len);
```
because both of these are used to parse the CCC:
```
int sc_asn1_read_tag(const u8 ** buf, size_t buflen, unsigned int *cla_out,
                    unsigned int *tag_out, size_t *taglen);
```
```
int sc_simpletlv_read_tag(u8 **buf, size_t buflen, u8 *tag_out, size_t *taglen);
```

I would suggest that the `sc_simpletlv` use the same `const u8 ** buf` but could be a separate PR.